sql/localsql/0003_migration.go

```diff
@@ -0,0 +1,192 @@
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+
+	"github.com/spacemeshos/post/initialization"
+
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/sql"
+)
+
+func New0003Migration(dataDir string, poetClients []PoetClient) *migration0003 {
+	return &migration0003{
+		dataDir:     dataDir,
+		poetClients: poetClients,
+	}
+}
+
+type migration0003 struct {
+	dataDir     string
+	poetClients []PoetClient
+}
+
+func (migration0003) Name() string {
+	return "add nipost builder state"
+}
+
+func (migration0003) Order() int {
+	return 3
+}
+
+func (m migration0003) Apply(db sql.Executor) error {
+	_, err := db.Exec("ALTER TABLE nipost RENAME TO challenge;", nil, nil)
+	if err != nil {
+		return fmt.Errorf("rename nipost table: %w", err)
+	}
+
+	_, err = db.Exec("ALTER TABLE challenge ADD COLUMN poet_proof_ref CHAR(32);", nil, nil)
+	if err != nil {
+		return fmt.Errorf("add poet_proof_ref column to challenge table: %w", err)
+	}
+
+	_, err = db.Exec("ALTER TABLE challenge ADD COLUMN poet_proof_membership VARCHAR;", nil, nil)
+	if err != nil {
+		return fmt.Errorf("add poet_proof_membership column to challenge table: %w", err)
+	}
+
+	if _, err := db.Exec(`CREATE TABLE poet_registration (
+	    id            CHAR(32) NOT NULL,
+	    hash          CHAR(32) NOT NULL,
+	    address       VARCHAR NOT NULL,
+	    round_id      VARCHAR NOT NULL,
+	    round_end     INT NOT NULL,
+
+	    PRIMARY KEY (id, address)
+	) WITHOUT ROWID;`, nil, nil); err != nil {
+		return fmt.Errorf("create poet_registration table: %w", err)
+	}
+
+	if _, err := db.Exec(`CREATE TABLE nipost (
+		id            CHAR(32) PRIMARY KEY,
+		post_nonce    UNSIGNED INT NOT NULL,
+		post_indices  VARCHAR NOT NULL,
+		post_pow      UNSIGNED LONG INT NOT NULL,
+
+		num_units UNSIGNED INT NOT NULL,
+		vrf_nonce UNSIGNED LONG INT NOT NULL,
+
+		poet_proof_membership VARCHAR NOT NULL,
+		poet_proof_ref        CHAR(32) NOT NULL,
+		labels_per_unit       UNSIGNED INT NOT NULL
+	) WITHOUT ROWID;`, nil, nil); err != nil {
+		return fmt.Errorf("create nipost table: %w", err)
+	}
+
+	return m.moveNipostStateToDb(db, m.dataDir)
+}
+
+func (m migration0003) moveNipostStateToDb(db sql.Executor, dataDir string) error {
+	state, err := loadBuilderState(dataDir)
+	switch {
+	case errors.Is(err, fs.ErrNotExist):
+		return nil // no state to move
+	case err != nil:
+		return fmt.Errorf("load nipost builder state: %w", err)
+	default:
+	}
+
+	meta, err := initialization.LoadMetadata(dataDir)
+	if err != nil {
+		return fmt.Errorf("load post metadata: %w", err)
+	}
+
+	if len(state.PoetRequests) == 0 {
+		return discardBuilderState(dataDir) // Phase 0: Submit PoET challenge to PoET services
+	}
+	// Phase 0 completed.
+	for _, req := range state.PoetRequests {
+		address, err := m.getAddress(req.PoetServiceID)
+		if err != nil {
+			return fmt.Errorf("get address for poet service id %x: %w", req.PoetServiceID.ServiceID, err)
+		}
```

Comment:
A failing migration should revert all changes that happened during the DB migration. On that note I probably need to not delete the migrated `*.bin` files but rather rename them (`*.bak`) and only delete them after the DB migration successfully completes (or leave them there in case we want a way to be able to downgrade if something goes wrong).

> What if we roll this out and then one of the poets crashes completely?

Ideally users can revert back to 1.2.6 and continue without issues. 

I agree we shouldn't fail if any request fails, but maybe only if ALL fail (the same way we do in NIPostBuilder). This however has the drawback that some PoETs might be dropped during migration that will then be online again during the cyclegap 😕.

> IMHO, to make this migration reliable we should first change the way poets are configured.

This will probably break things. Changes to the config are not trivial in the setup we are currently using, since the config is overwritten by Smapp from the config here: https://configs.spacemesh.network/config.mainnet.json So any change we make must work with both old and new nodes, i.e. we need to have poets twice in the config (old and new layout).

I think maybe we should just hard code a map of the official PoETs (Key -> URL) and provide a way for users to provide such a mapping for their custom PoETs if needed; but not via the config but a new file that is only read during migration and can be discarded afterwards.

Reply:
By "poet crashes completely" I meant a situation when one of the poets dies and doesn't boot anymore, and needs maintenance. In such a situation, all nodes would fail the migration until we fix the poet (if we even can fix it - it could take a week or so). In effect, the upgrade would be blocked until we fix the poet, which IMHO is a no-go.

---

>>   IMHO, to make this migration reliable we should first change the way poets are configured.
>
> This will probably break things. Changes to the config are not trivial in the setup we are currently using, since the config is overwritten by Smapp from the config here: https://configs.spacemesh.network/config.mainnet.json So any change we make must work with both old and new nodes, i.e. we need to have poets twice in the config (old and new layout).

Are you sure? Can't Smapp use config **A** for node version **a** and config **B** for node version **b**? Even if so, I think having poets temporarily configured twice (two separate fields) in the ` https://configs.spacemesh.network/config.mainnet.json` is not a big deal. It doesn't mean that the config structure of the new node version needs both. It will only contain the new field and the older one should be omitted during parsing.

We should communicate upgrade steps properly in the release nodes with an instruction on how to migrate the config in case somebody overwrites the list of poet servers (the default config would contain the list of SM poets with their pubkeys).

I can create a PR to change poet configuration as described.

Reply:
I see three options:

- fail the migration if a PoET isn't available while the migration happens. This will revert all changes and users can try again or users can downgrade to v1.2.6 and try upgrading later. As you said that's probably not the best solution since worst case scenario people might to wait a full epoch to be able to upgrade (then the PoET wouldn't be in the new `nipost_builder_state.bin` file any more if it is still down)
- not fail the migration if at least one PoET is available when the migration happens. This would match the behaviour that the `NIPostBuilder` uses to fetch PoET proofs and would be my preferred solution. A PoET that is down (error or maintenance) is unlikely to have the highest tick count anyway, so losing that registration might not even matter.
- change the config to include both the URL and the public key of a PoET and force users to update their config accordingly. I would like to avoid that change, people might use the wrong public keys for the nodes (copy & paste errors, typos, swap keys accidentally) which can lead to multiple problems during migration.

> Can't Smapp use config A for node version a and config B for node version b?

Afaik at the moment it cannot; also a lot of users don't use Smapp.

> We should communicate upgrade steps properly in the release nodes with an instruction on how to migrate the config in case somebody overwrites the list of poet servers (the default config would contain the list of SM poets with their pubkeys).

That could work if you think that the automatic migration with at least one PoET successfully migrated is too risky.

Wdyt?

Reply:
I finished and opened PR allowing to configure poets' public keys: https://github.com/spacemeshos/go-spacemesh/pull/5293.

Regarding:
>>Can't Smapp use config A for node version a and config B for node version b?
>
>Afaik at the moment it cannot; also a lot of users don't use Smapp.

@nj confirmed that it can serve different configs based on node version.

Reply:
@poszu Updated the PR after #5293 was merged. It now uses the public keys from the config instead of querying them.